src/headerssync.cpp

```diff
@@ -0,0 +1,318 @@
+            // The headers didn't pass validation; give up on the sync.
+            return std::nullopt;
+        }
+        processing_success = true;
+        if (full_headers_message || m_download_state == State::REDOWNLOAD) {
+            // A full headers message means the peer may have more to give us;
+            // also if we just switched to REDOWNLOAD then we need to re-request
+            // headers from the beginning.
+            return MakeNextHeadersRequest();
+        } else {
+            // If we're in INITIAL_DOWNLOAD and we get a non-full headers
+            // message, then the peer's chain has ended and definitely doesn't
+            // have enough work, so we can stop our sync.
+            Finalize();
+            return std::nullopt;
+        }
+    } else if (m_download_state == State::REDOWNLOAD) {
+        // During REDOWNLOAD, we compare our stored commitments to what we
+        // receive, and add headers to our redownload buffer. When the buffer
+        // gets big enough (meaning that we've checked enough commitments),
+        // we'll return a batch of headers to the caller for processing.
+        for (const auto& hdr : headers) {
+            if (!ValidateAndStoreRedownloadedHeader(hdr)) {
+                // Something went wrong -- the peer gave us an unexpected chain.
+                // We could consider looking at the reason for failure and
+                // punishing the peer, but for now just give up on sync.
+                Finalize();
+                return std::nullopt;
+            }
+        }
+
+        processing_success = true;
+        // Return any headers that are ready for acceptance.
+        headers_to_process = RemoveHeadersReadyForAcceptance();
+
+        // If we hit our target blockhash, then all remaining headers will be
+        // returned and we can clear any leftover internal state.
+        if (m_redownloaded_headers.empty() && m_process_all_remaining_headers) {
+            Finalize();
+            return std::nullopt;
+        }
+
+        // If the headers message is full, we need to request more.
+        if (full_headers_message) {
+            return MakeNextHeadersRequest();
+        } else {
+            // For some reason our peer gave us a high-work chain, but is now
+            // declining to serve us that full chain again. Give up.
+            // Note that there's no more processing to be done with these
+            // headers, so we can still return success.
+            Finalize();
+            return std::nullopt;
+        }
+    }
+    return std::nullopt;
+}
+
+bool HeadersSyncState::ValidateAndStoreHeadersCommitments(const std::vector<CBlockHeader>& headers)
+{
+    // The caller should not give us an empty set of headers.
+    Assume(headers.size() > 0);
+    if (headers.size() == 0) return true;
+
+    Assume(m_download_state != State::FINAL);
+    if (m_download_state == State::FINAL) return false;
+
+    if (m_last_header_received.IsNull()) {
```

Comment:
In commit "Utilize anti-DoS headers download strategy"

Any reason to do this here, and not in `StartInitialDownload`? It seems a bit cleaner to set these values as soon as they're known, rather than on the fly.

Reply:
Done.